### ISSUE TITLE
Splitting mesh seems slow

### ISSUE THREAD

**Issue by YaqiWang** [libMesh/libmesh#1736]
I am just running a case with 262983 elements with 4608 cores. I know I am over using number of cores but it should be irrelevant. The mesh is in ExodusII format and not pre-split. When I use distributed mesh, I can see it takes quite a long time (more than few minutes) to finish construction of the mesh, although I am expecting seconds. Sorry I cannot provide the profiling information, but to looks me that the code somewhere can be improved.

**Comment by jwpeterson** [comment]
> The mesh is in ExodusII format and not pre-split.

If the Mesh is not pre-split, then I don't think you reap the full benefits of using `DistributedMesh`. The time you are seeing may be the time required to read the Mesh on processor 0 and communicate the relevant pieces (this may even be a broadcast?) to the rest of the procs. I could definitely see this being a bottleneck.

**Comment by jwpeterson** [comment]
The function being called is probably `ExodusII_IO::read()` in `libmesh/src/mesh/exodusII_io.C`. In looking over the code, it actually seems like the Mesh file is opened and read (in its entirety) on *every* proc. Only then are the remote elements discarded. I don't think this is going to be usable at scale at all, hence the reason the pre-split capability exists...

**Comment by YaqiWang** [comment]
Let me check if that that it is. Should we do broadcast then?

**Comment by jwpeterson** [comment]
> Should we do broadcast then?

This would be a big code change and might still be slow. Is there a reason you are not using the pre-split approach? Did it fail or something?

**Comment by permcody** [comment]
> The function being called is probably ExodusII_IO::read() in libmesh/src/mesh/exodusII_io.C.

Wait, are you sure? If we read in the mesh we normally do it through UnstructuredMesh::read(), which definitely only reads it on processor zero. The exception is if you are explicitly doing one of the initial conditions that explicitly calls the Exodus read function directly.

**Comment by YaqiWang** [comment]
No, just being lazy to issue two commands and generate all the splits, which I am not so sure while playing with the code. Pre-splitting is slow too probably due to the same reason.

**Comment by jwpeterson** [comment]
> Wait, are you sure? If we read in the mesh we normally do it through UnstructuredMesh::read(), which definitely only reads it on processor zero.

Ah, yeah the `NameBasedIO` stuff does do this, followed by a broadcast.

**Comment by roystgnr** [comment]
> The exception is if you are explicitly doing one of the initial conditions that explicitly calls the Exodus read function directly.

Should that be necessary?  It should still be more efficient to read on one processor and then broadcast, right?

**Comment by YaqiWang** [comment]
ExodusII_IO::read() seems indeed being called and done fast at least on processor 0. I can do a MPI barrier to make sure all processors hit the printout at the same time.

**Comment by permcody** [comment]
> Should that be necessary? It should still be more efficient to read on one processor and then broadcast, right?

Not necessary. Somebody just needs to make it work!

**Comment by YaqiWang** [comment]
I was stupid, MOOSE show_actions capability actually tallied the time of all actions, and it shows that 
```
|   Read Mesh                                      1          204.4752    204.475154  204.4752    204.475154  9.13     9.13     |
```
which is clearly indicate that reading mesh is slow. 
Preparing mesh takes
```
|   prepare_mesh:Mesh                              1          11.7321     11.732056   11.7321     11.732056   0.52     0.52     |
```
So I guess the request of this issue will be loading exodus mesh only on master followed by a broadcast.

**Comment by permcody** [comment]
Yaqi, we really do only read on one processor. Take a look at `UnstructuredMesh::read()`

**Comment by YaqiWang** [comment]
Right. I was ignoring UnstructuredMesh::read(). This also explain why I cannot put barrier inside ExodusII_IO::read(). I have confirmed that this line is the culprit: https://github.com/libMesh/libmesh/blob/master/src/mesh/unstructured_mesh.C#L621

Time in reading mesh on master:2.45949s
Time in broadcast:2.01766s
Time in prepare_for_use:198.156s

Reading file and broadcast are fast as expected.

**Comment by YaqiWang** [comment]
This is the break-down of cpu time in prepare_for_use:

First call to renumber_nodes_and_elements():  38.3023s
Call to partition(): 161.516s
Second call to renumber_nodes_and_elements(): 2.19678s

I am using parmetis for partitioning. Looks like this is not a libMesh issue. If you think the time in renumber_nodes_and_elements() is reasonable, feel free to close this. It is surprising to me that the partitioner can take this much time...

**Comment by jwpeterson** [comment]
Seems like a big disparity between the first and second calls to `renumber_nodes_and_elements()` but maybe there is less work to do the second time around? 

As far as the partition time, we always call `ParMETIS_V3_AdaptiveRepart()` in libmesh, but perhaps that algorithm is not the fastest for the number of processors you are currently running on...

**Comment by YaqiWang** [comment]
The second call is after partitioning, so I guess too there is less work in it. @friedmud talked to me that parmetis itself has too many communications in this setting (262983 elements with 4608 cores). He suspects metis can out-perform parmetis here. I am happy this is not in libMesh, not like what I thought.

**Comment by roystgnr** [comment]
> I am using parmetis for partitioning. Looks like this is not a libMesh issue.

It's looking like this *was* a libMesh issue, in the setup for ParMETIS.  After we get #2653 merged, check and see if that sped up splitting for you?
